// @waku/core/dist/lib/keep_alive_manager.js
import debug from "debug";
import { createEncoder } from "../index.js";
import { RelayPingContentTopic } from "./relay/constants.js";
const log = debug("waku:keep-alive");
export class KeepAliveManager {
    constructor(options, relay) {
        this.pingKeepAliveTimers = new Map();
        this.relayKeepAliveTimers = new Map();
        this.options = options;
        this.relay = relay;
    }
    start(peerId, libp2pPing) {
        // Just in case a timer already exist for this peer
        this.stop(peerId);
        const { pingKeepAlive: pingPeriodSecs, relayKeepAlive: relayPeriodSecs } = this.options;
        const peerIdStr = peerId.toString();
        if (pingPeriodSecs !== 0) {
            const interval = setInterval(() => {
                libp2pPing(peerId).catch((e) => {
                    log(`Ping failed (${peerIdStr})`, e);
                });
            }, pingPeriodSecs * 1000);
            this.pingKeepAliveTimers.set(peerIdStr, interval);
        }
        const relay = this.relay;
        if (relay && relayPeriodSecs !== 0) {
            const encoder = createEncoder({
                contentTopic: RelayPingContentTopic,
                ephemeral: true,
            });
            const interval = setInterval(() => {
                log("Sending Waku Relay ping message");
                relay
                    .send(encoder, { payload: new Uint8Array([1]) })
                    .catch((e) => log("Failed to send relay ping", e));
            }, relayPeriodSecs * 1000);
            this.relayKeepAliveTimers.set(peerId, interval);
        }
    }
    stop(peerId) {
        const peerIdStr = peerId.toString();
        if (this.pingKeepAliveTimers.has(peerIdStr)) {
            clearInterval(this.pingKeepAliveTimers.get(peerIdStr));
            this.pingKeepAliveTimers.delete(peerIdStr);
        }
        if (this.relayKeepAliveTimers.has(peerId)) {
            clearInterval(this.relayKeepAliveTimers.get(peerId));
            this.relayKeepAliveTimers.delete(peerId);
        }
    }
    stopAll() {
        for (const timer of [
            ...Object.values(this.pingKeepAliveTimers),
            ...Object.values(this.relayKeepAliveTimers),
        ]) {
            clearInterval(timer);
        }
        this.pingKeepAliveTimers.clear();
        this.relayKeepAliveTimers.clear();
    }
}
