// @waku/core/bundle/topic_only_message-e8406994.js
import { T as TopicOnlyMessage$1 } from './peer_exchange-1229c8b0.js';
import { d as debug } from './browser-2f1afe46.js';

const log = debug("waku:message:topic-only");
class TopicOnlyMessage {
    constructor(pubSubTopic, proto) {
        this.pubSubTopic = pubSubTopic;
        this.proto = proto;
        this.payload = new Uint8Array();
    }
    get contentTopic() {
        return this.proto.contentTopic;
    }
}
class TopicOnlyDecoder {
    constructor() {
        this.contentTopic = "";
    }
    fromWireToProtoObj(bytes) {
        const protoMessage = TopicOnlyMessage$1.decode(bytes);
        log("Message decoded", protoMessage);
        return Promise.resolve({
            contentTopic: protoMessage.contentTopic,
            payload: new Uint8Array(),
            rateLimitProof: undefined,
            timestamp: undefined,
            meta: undefined,
            version: undefined,
            ephemeral: undefined,
        });
    }
    async fromProtoObj(pubSubTopic, proto) {
        return new TopicOnlyMessage(pubSubTopic, proto);
    }
}

var topic_only_message = /*#__PURE__*/Object.freeze({
    __proto__: null,
    TopicOnlyDecoder: TopicOnlyDecoder,
    TopicOnlyMessage: TopicOnlyMessage
});

export { TopicOnlyDecoder as T, TopicOnlyMessage as a, topic_only_message as t };
