// @waku/dns-discovery/dist/index.js
import { symbol } from "@libp2p/interface-peer-discovery";
import { CustomEvent, EventEmitter } from "@libp2p/interfaces/events";
import debug from "debug";
import { DnsNodeDiscovery } from "./dns.js";
const log = debug("waku:peer-discovery-dns");
const enrTree = {
    TEST: "enrtree://AOGECG2SPND25EEFMAJ5WF3KSGJNSGV356DSTL2YVLLZWIV6SAYBM@test.waku.nodes.status.im",
    PROD: "enrtree://AOGECG2SPND25EEFMAJ5WF3KSGJNSGV356DSTL2YVLLZWIV6SAYBM@prod.waku.nodes.status.im",
};
const DEFAULT_BOOTSTRAP_TAG_NAME = "bootstrap";
const DEFAULT_BOOTSTRAP_TAG_VALUE = 50;
const DEFAULT_BOOTSTRAP_TAG_TTL = 120000;
/**
 * Parse options and expose function to return bootstrap peer addresses.
 */
export class PeerDiscoveryDns extends EventEmitter {
    constructor(components, options) {
        super();
        this._started = false;
        this._components = components;
        this._options = options;
        const { enrUrl } = options;
        log("Use following EIP-1459 ENR Tree URL: ", enrUrl);
    }
    /**
     * Start discovery process
     */
    async start() {
        log("Starting peer discovery via dns");
        this._started = true;
        if (this.nextPeer === undefined) {
            const { enrUrl, wantedNodeCapabilityCount } = this._options;
            const dns = await DnsNodeDiscovery.dnsOverHttp();
            this.nextPeer = dns.getNextPeer.bind(dns, [enrUrl], wantedNodeCapabilityCount);
        }
        for await (const peer of this.nextPeer()) {
            if (!this._started)
                return;
            const peerInfo = peer.peerInfo;
            if (!peerInfo)
                continue;
            if ((await this._components.peerStore.getTags(peerInfo.id)).find(({ name }) => name === DEFAULT_BOOTSTRAP_TAG_NAME))
                continue;
            await this._components.peerStore.tagPeer(peerInfo.id, DEFAULT_BOOTSTRAP_TAG_NAME, {
                value: this._options.tagValue ?? DEFAULT_BOOTSTRAP_TAG_VALUE,
                ttl: this._options.tagTTL ?? DEFAULT_BOOTSTRAP_TAG_TTL,
            });
            this.dispatchEvent(new CustomEvent("peer", { detail: peerInfo }));
        }
    }
    /**
     * Stop emitting events
     */
    stop() {
        this._started = false;
    }
    get [symbol]() {
        return true;
    }
    get [Symbol.toStringTag]() {
        return "@waku/bootstrap";
    }
}
export function wakuDnsDiscovery(enrUrl, wantedNodeCapabilityCount) {
    return (components) => new PeerDiscoveryDns(components, { enrUrl, wantedNodeCapabilityCount });
}
export { DnsNodeDiscovery } from "./dns.js";
export { enrTree };
