// greenwebjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Singleton = exports.CAT = exports.StandardCoin = exports.SmartCoin = exports.CoinSpend = exports.Coin = exports.BigNumber = exports.spend = exports.util = exports.clvm = exports.xch = void 0;
// greenweb.xch
var xch_1 = require("./xch");
Object.defineProperty(exports, "xch", { enumerable: true, get: function () { return xch_1.XCHModule; } });
// greenweb.clvm
exports.clvm = __importStar(require("clvm"));
// greenweb.util
var util_1 = require("./util");
Object.defineProperty(exports, "util", { enumerable: true, get: function () { return util_1.Util; } });
// greenweb.spend
var spend_1 = require("./spend");
Object.defineProperty(exports, "spend", { enumerable: true, get: function () { return spend_1.SpendModule; } });
// greenweb.BigNumber
var bignumber_1 = require("@ethersproject/bignumber");
Object.defineProperty(exports, "BigNumber", { enumerable: true, get: function () { return bignumber_1.BigNumber; } });
// greenweb.Coin
var coin_1 = require("./util/serializer/types/coin");
Object.defineProperty(exports, "Coin", { enumerable: true, get: function () { return coin_1.Coin; } });
// greenWeb.CoinSpend
var coin_spend_1 = require("./util/serializer/types/coin_spend");
Object.defineProperty(exports, "CoinSpend", { enumerable: true, get: function () { return coin_spend_1.CoinSpend; } });
// greenweb.SmartCoin
var smart_coin_1 = require("./smart_coin");
Object.defineProperty(exports, "SmartCoin", { enumerable: true, get: function () { return smart_coin_1.SmartCoin; } });
// greenweb.StandardCoin
var standard_coin_1 = require("./standard_coin");
Object.defineProperty(exports, "StandardCoin", { enumerable: true, get: function () { return standard_coin_1.StandardCoin; } });
// greenweb.CAT
var cat_1 = require("./cat");
Object.defineProperty(exports, "CAT", { enumerable: true, get: function () { return cat_1.CAT; } });
// greenweb.Singleton
var singleton_1 = require("./singleton");
Object.defineProperty(exports, "Singleton", { enumerable: true, get: function () { return singleton_1.Singleton; } });
