// greenwebjs/util/serializer/types/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const coin_1 = require("./coin");
const coin_spend_1 = require("./coin_spend");
const end_of_sub_slot_bundle_1 = require("./end_of_sub_slot_bundle");
const fields_1 = __importDefault(require("./fields"));
const foliage_1 = require("./foliage");
const header_block_1 = require("./header_block");
const outbound_message_1 = require("./outbound_message");
const pool_target_1 = require("./pool_target");
const proof_of_space_1 = require("./proof_of_space");
const reward_chain_block_1 = require("./reward_chain_block");
const shared_protocol_1 = require("./shared_protocol");
const slots_1 = require("./slots");
const spend_bundle_1 = require("./spend_bundle");
const vdf_1 = require("./vdf");
const wallet_protocol_1 = require("./wallet_protocol");
exports.default = {
    fields: fields_1.default,
    CoinSpend: coin_spend_1.CoinSpend,
    Coin: coin_1.Coin,
    EndOfSubSlotBundle: end_of_sub_slot_bundle_1.EndOfSubSlotBundle,
    TransactionsInfo: foliage_1.TransactionsInfo,
    FoliageTransactionBlock: foliage_1.FoliageTransactionBlock,
    FoliageBlockData: foliage_1.FoliageBlockData,
    Foliage: foliage_1.Foliage,
    HeaderBlock: header_block_1.HeaderBlock,
    Message: outbound_message_1.Message,
    PoolTarget: pool_target_1.PoolTarget,
    ProofOfSpace: proof_of_space_1.ProofOfSpace,
    RewardChainBlock: reward_chain_block_1.RewardChainBlock,
    Handshake: shared_protocol_1.Handshake,
    ChallengeBlockInfo: slots_1.ChallengeBlockInfo,
    ChallengeChainSubSlot: slots_1.ChallengeChainSubSlot,
    InfusedChallengeChainSubSlot: slots_1.InfusedChallengeChainSubSlot,
    RewardChainSubSlot: slots_1.RewardChainSubSlot,
    SubSlotProofs: slots_1.SubSlotProofs,
    SpendBundle: spend_bundle_1.SpendBundle,
    VDFInfo: vdf_1.VDFInfo,
    VDFProof: vdf_1.VDFProof,
    RequestPuzzleSolution: wallet_protocol_1.RequestPuzzleSolution,
    PuzzleSolutionResponse: wallet_protocol_1.PuzzleSolutionResponse,
    RespondPuzzleSolution: wallet_protocol_1.RespondPuzzleSolution,
    RejectPuzzleSolution: wallet_protocol_1.RejectPuzzleSolution,
    SendTransaction: wallet_protocol_1.SendTransaction,
    TransactionAck: wallet_protocol_1.TransactionAck,
    NewPeakWallet: wallet_protocol_1.NewPeakWallet,
    RequestBlockHeader: wallet_protocol_1.RequestBlockHeader,
    RespondBlockHeader: wallet_protocol_1.RespondBlockHeader,
    RejectHeaderRequest: wallet_protocol_1.RejectHeaderRequest,
    RequestRemovals: wallet_protocol_1.RequestRemovals,
    RespondRemovals: wallet_protocol_1.RespondRemovals,
    RejectRemovalsRequest: wallet_protocol_1.RejectRemovalsRequest,
    RequestAdditions: wallet_protocol_1.RequestAdditions,
    RespondAdditions: wallet_protocol_1.RespondAdditions,
    RejectAdditionsRequest: wallet_protocol_1.RejectAdditionsRequest,
    RequestHeaderBlocks: wallet_protocol_1.RequestHeaderBlocks,
    RejectHeaderBlocks: wallet_protocol_1.RejectHeaderBlocks,
    RespondHeaderBlocks: wallet_protocol_1.RespondHeaderBlocks,
    CoinState: wallet_protocol_1.CoinState,
    RegisterForPhUpdates: wallet_protocol_1.RegisterForPhUpdates,
    RespondToPhUpdates: wallet_protocol_1.RespondToPhUpdates,
    RegisterForCoinUpdates: wallet_protocol_1.RegisterForCoinUpdates,
    RespondToCoinUpdates: wallet_protocol_1.RespondToCoinUpdates,
    CoinStateUpdate: wallet_protocol_1.CoinStateUpdate,
    RequestChildren: wallet_protocol_1.RequestChildren,
    RespondChildren: wallet_protocol_1.RespondChildren,
    RequestSESInfo: wallet_protocol_1.RequestSESInfo,
    RespondSESInfo: wallet_protocol_1.RespondSESInfo
};
