// greenwebjs/util/serializer/types/protocol_message_types.js
"use strict";
// https://github.com/Chia-Network/chia-blockchain/blob/main/chia/protocols/protocol_message_types.py#L4
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtocolMessageTypes = void 0;
var ProtocolMessageTypes;
(function (ProtocolMessageTypes) {
    ProtocolMessageTypes[ProtocolMessageTypes["handshake"] = 1] = "handshake";
    // Harvester protocol (harvester <-> farmer)
    ProtocolMessageTypes[ProtocolMessageTypes["harvester_handshake"] = 3] = "harvester_handshake";
    // new_signage_point_harvester = 4 Changed to 66 in new protocol
    ProtocolMessageTypes[ProtocolMessageTypes["new_proof_of_space"] = 5] = "new_proof_of_space";
    ProtocolMessageTypes[ProtocolMessageTypes["request_signatures"] = 6] = "request_signatures";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_signatures"] = 7] = "respond_signatures";
    // Farmer protocol (farmer <-> full_node)
    ProtocolMessageTypes[ProtocolMessageTypes["new_signage_point"] = 8] = "new_signage_point";
    ProtocolMessageTypes[ProtocolMessageTypes["declare_proof_of_space"] = 9] = "declare_proof_of_space";
    ProtocolMessageTypes[ProtocolMessageTypes["request_signed_values"] = 10] = "request_signed_values";
    ProtocolMessageTypes[ProtocolMessageTypes["signed_values"] = 11] = "signed_values";
    ProtocolMessageTypes[ProtocolMessageTypes["farming_info"] = 12] = "farming_info";
    // Timelord protocol (timelord <-> full_node)
    ProtocolMessageTypes[ProtocolMessageTypes["new_peak_timelord"] = 13] = "new_peak_timelord";
    ProtocolMessageTypes[ProtocolMessageTypes["new_unfinished_block_timelord"] = 14] = "new_unfinished_block_timelord";
    ProtocolMessageTypes[ProtocolMessageTypes["new_infusion_point_vdf"] = 15] = "new_infusion_point_vdf";
    ProtocolMessageTypes[ProtocolMessageTypes["new_signage_point_vdf"] = 16] = "new_signage_point_vdf";
    ProtocolMessageTypes[ProtocolMessageTypes["new_end_of_sub_slot_vdf"] = 17] = "new_end_of_sub_slot_vdf";
    ProtocolMessageTypes[ProtocolMessageTypes["request_compact_proof_of_time"] = 18] = "request_compact_proof_of_time";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_compact_proof_of_time"] = 19] = "respond_compact_proof_of_time";
    // Full node protocol (full_node <-> full_node)
    ProtocolMessageTypes[ProtocolMessageTypes["new_peak"] = 20] = "new_peak";
    ProtocolMessageTypes[ProtocolMessageTypes["new_transaction"] = 21] = "new_transaction";
    ProtocolMessageTypes[ProtocolMessageTypes["request_transaction"] = 22] = "request_transaction";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_transaction"] = 23] = "respond_transaction";
    ProtocolMessageTypes[ProtocolMessageTypes["request_proof_of_weight"] = 24] = "request_proof_of_weight";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_proof_of_weight"] = 25] = "respond_proof_of_weight";
    ProtocolMessageTypes[ProtocolMessageTypes["request_block"] = 26] = "request_block";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_block"] = 27] = "respond_block";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_block"] = 28] = "reject_block";
    ProtocolMessageTypes[ProtocolMessageTypes["request_blocks"] = 29] = "request_blocks";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_blocks"] = 30] = "respond_blocks";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_blocks"] = 31] = "reject_blocks";
    ProtocolMessageTypes[ProtocolMessageTypes["new_unfinished_block"] = 32] = "new_unfinished_block";
    ProtocolMessageTypes[ProtocolMessageTypes["request_unfinished_block"] = 33] = "request_unfinished_block";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_unfinished_block"] = 34] = "respond_unfinished_block";
    ProtocolMessageTypes[ProtocolMessageTypes["new_signage_point_or_end_of_sub_slot"] = 35] = "new_signage_point_or_end_of_sub_slot";
    ProtocolMessageTypes[ProtocolMessageTypes["request_signage_point_or_end_of_sub_slot"] = 36] = "request_signage_point_or_end_of_sub_slot";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_signage_point"] = 37] = "respond_signage_point";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_end_of_sub_slot"] = 38] = "respond_end_of_sub_slot";
    ProtocolMessageTypes[ProtocolMessageTypes["request_mempool_transactions"] = 39] = "request_mempool_transactions";
    ProtocolMessageTypes[ProtocolMessageTypes["request_compact_vdf"] = 40] = "request_compact_vdf";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_compact_vdf"] = 41] = "respond_compact_vdf";
    ProtocolMessageTypes[ProtocolMessageTypes["new_compact_vdf"] = 42] = "new_compact_vdf";
    ProtocolMessageTypes[ProtocolMessageTypes["request_peers"] = 43] = "request_peers";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_peers"] = 44] = "respond_peers";
    // Wallet protocol (wallet <-> full_node)
    ProtocolMessageTypes[ProtocolMessageTypes["request_puzzle_solution"] = 45] = "request_puzzle_solution";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_puzzle_solution"] = 46] = "respond_puzzle_solution";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_puzzle_solution"] = 47] = "reject_puzzle_solution";
    ProtocolMessageTypes[ProtocolMessageTypes["send_transaction"] = 48] = "send_transaction";
    ProtocolMessageTypes[ProtocolMessageTypes["transaction_ack"] = 49] = "transaction_ack";
    ProtocolMessageTypes[ProtocolMessageTypes["new_peak_wallet"] = 50] = "new_peak_wallet";
    ProtocolMessageTypes[ProtocolMessageTypes["request_block_header"] = 51] = "request_block_header";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_block_header"] = 52] = "respond_block_header";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_header_request"] = 53] = "reject_header_request";
    ProtocolMessageTypes[ProtocolMessageTypes["request_removals"] = 54] = "request_removals";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_removals"] = 55] = "respond_removals";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_removals_request"] = 56] = "reject_removals_request";
    ProtocolMessageTypes[ProtocolMessageTypes["request_additions"] = 57] = "request_additions";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_additions"] = 58] = "respond_additions";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_additions_request"] = 59] = "reject_additions_request";
    ProtocolMessageTypes[ProtocolMessageTypes["request_header_blocks"] = 60] = "request_header_blocks";
    ProtocolMessageTypes[ProtocolMessageTypes["reject_header_blocks"] = 61] = "reject_header_blocks";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_header_blocks"] = 62] = "respond_header_blocks";
    // Introducer protocol (introducer <-> full_node)
    ProtocolMessageTypes[ProtocolMessageTypes["request_peers_introducer"] = 63] = "request_peers_introducer";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_peers_introducer"] = 64] = "respond_peers_introducer";
    // Simulator protocol
    ProtocolMessageTypes[ProtocolMessageTypes["farm_new_block"] = 65] = "farm_new_block";
    // New harvester protocol
    ProtocolMessageTypes[ProtocolMessageTypes["new_signage_point_harvester"] = 66] = "new_signage_point_harvester";
    ProtocolMessageTypes[ProtocolMessageTypes["request_plots"] = 67] = "request_plots";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_plots"] = 68] = "respond_plots";
    // More wallet protocol
    ProtocolMessageTypes[ProtocolMessageTypes["coin_state_update"] = 69] = "coin_state_update";
    ProtocolMessageTypes[ProtocolMessageTypes["register_interest_in_puzzle_hash"] = 70] = "register_interest_in_puzzle_hash";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_to_ph_update"] = 71] = "respond_to_ph_update";
    ProtocolMessageTypes[ProtocolMessageTypes["register_interest_in_coin"] = 72] = "register_interest_in_coin";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_to_coin_update"] = 73] = "respond_to_coin_update";
    ProtocolMessageTypes[ProtocolMessageTypes["request_children"] = 74] = "request_children";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_children"] = 75] = "respond_children";
    ProtocolMessageTypes[ProtocolMessageTypes["request_ses_hashes"] = 76] = "request_ses_hashes";
    ProtocolMessageTypes[ProtocolMessageTypes["respond_ses_hashes"] = 77] = "respond_ses_hashes";
})(ProtocolMessageTypes = exports.ProtocolMessageTypes || (exports.ProtocolMessageTypes = {}));
