// @waku/core/dist/lib/filter/filter_rpc.js
import { proto_filter as proto } from "@waku/proto";
import { v4 as uuid } from "uuid";
/**
 * FilterRPC represents a message conforming to the Waku Filter protocol
 */
export class FilterRpc {
    constructor(proto) {
        this.proto = proto;
    }
    static createRequest(topic, contentFilters, requestId, subscribe = true) {
        return new FilterRpc({
            requestId: requestId || uuid(),
            request: {
                subscribe,
                topic,
                contentFilters,
            },
            push: undefined,
        });
    }
    /**
     *
     * @param bytes Uint8Array of bytes from a FilterRPC message
     * @returns FilterRpc
     */
    static decode(bytes) {
        const res = proto.FilterRpc.decode(bytes);
        return new FilterRpc(res);
    }
    /**
     * Encode the current FilterRPC request to bytes
     * @returns Uint8Array
     */
    encode() {
        return proto.FilterRpc.encode(this.proto);
    }
    get push() {
        return this.proto.push;
    }
    get requestId() {
        return this.proto.requestId;
    }
}
