// greenwebjs/util/sexp/condition_opcodes.js
"use strict";
// https://github.com/Chia-Network/chia-blockchain/blob/d3e73a75ab44c799f8b6f9a76fab550ad6d7824a/chia/types/condition_opcodes.py#L6
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConditionOpcode = void 0;
var ConditionOpcode;
(function (ConditionOpcode) {
    // AGG_SIG is ascii "1"
    // the conditions below require bls12-381 signatures
    ConditionOpcode["AGG_SIG_UNSAFE"] = "31";
    ConditionOpcode["AGG_SIG_ME"] = "32";
    // the conditions below reserve coin amounts and have to be accounted for in output totals
    ConditionOpcode["CREATE_COIN"] = "33";
    ConditionOpcode["RESERVE_FEE"] = "34";
    // the conditions below deal with announcements, for inter-coin communication
    ConditionOpcode["CREATE_COIN_ANNOUNCEMENT"] = "3c";
    ConditionOpcode["ASSERT_COIN_ANNOUNCEMENT"] = "3d";
    ConditionOpcode["CREATE_PUZZLE_ANNOUNCEMENT"] = "3e";
    ConditionOpcode["ASSERT_PUZZLE_ANNOUNCEMENT"] = "3f";
    // the conditions below let coins inquire about themselves
    ConditionOpcode["ASSERT_MY_COIN_ID"] = "46";
    ConditionOpcode["ASSERT_MY_PARENT_ID"] = "47";
    ConditionOpcode["ASSERT_MY_PUZZLEHASH"] = "48";
    ConditionOpcode["ASSERT_MY_AMOUNT"] = "49";
    // the conditions below ensure that we're "far enough" in the future
    // wall-clock time
    ConditionOpcode["ASSERT_SECONDS_RELATIVE"] = "50";
    ConditionOpcode["ASSERT_SECONDS_ABSOLUTE"] = "51";
    // block index
    ConditionOpcode["ASSERT_HEIGHT_RELATIVE"] = "52";
    ConditionOpcode["ASSERT_HEIGHT_ABSOLUTE"] = "53";
})(ConditionOpcode = exports.ConditionOpcode || (exports.ConditionOpcode = {}));
