// @waku/core/dist/lib/message/topic_only_message.js
import { TopicOnlyMessage as ProtoTopicOnlyMessage } from "@waku/proto";
import debug from "debug";
const log = debug("waku:message:topic-only");
export class TopicOnlyMessage {
    constructor(pubSubTopic, proto) {
        this.pubSubTopic = pubSubTopic;
        this.proto = proto;
        this.payload = new Uint8Array();
    }
    get contentTopic() {
        return this.proto.contentTopic;
    }
}
export class TopicOnlyDecoder {
    constructor() {
        this.contentTopic = "";
    }
    fromWireToProtoObj(bytes) {
        const protoMessage = ProtoTopicOnlyMessage.decode(bytes);
        log("Message decoded", protoMessage);
        return Promise.resolve({
            contentTopic: protoMessage.contentTopic,
            payload: new Uint8Array(),
            rateLimitProof: undefined,
            timestamp: undefined,
            meta: undefined,
            version: undefined,
            ephemeral: undefined,
        });
    }
    async fromProtoObj(pubSubTopic, proto) {
        return new TopicOnlyMessage(pubSubTopic, proto);
    }
}
